### PR TITLE
[Prost] Remove extra generated modules and opt-in to transitively including deps

### PR DESCRIPTION
**Pull request by freeformstu**
This PR makes two changes:
1. It removes the duplicately generated code in the prost `lib.rs`.
2. It allows opting into transitively including all of the proto deps.

## Duplicately generated code.

Previously there was a bug where we generated the same code at every proto package level. This created separate Rust types for each module which made it easy to accidentally import an incompatible type.

Before:
```rust
// @generated

pub mod b_ar {
    pub mod b_az {
        pub mod qaz {
            pub mod qu_x {
                // @generated
                // This file is @generated by prost-build.
                #[allow(clippy::derive_partial_eq_without_eq)]
                #[derive(Clone, PartialEq, ::prost::Message)]
                pub struct Bar {
                    #[prost(string, tag = "1")]
                    pub name: ::prost::alloc::string::String,
                    #[prost(message, optional, tag = "2")]
                    pub foo: ::core::option::Option<
                        ::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::Foo,
                    >,
                    #[prost(message, optional, tag = "3")]
                    pub nested_foo: ::core::option::Option<
                        ::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::foo::NestedFoo,
                    >,
                }
                /// Nested message and enum types in `Bar`.
                pub mod bar {
                    #[allow(clippy::derive_partial_eq_without_eq)]
                    #[derive(Clone, PartialEq, ::prost::Message)]
                    pub struct Baz {
                        #[prost(string, tag = "4")]
                        pub name: ::prost::alloc::string::String,
                    }
                }
                // @@protoc_insertion_point(module)
            }
        }
    }
}
pub mod b_az {
    pub mod qaz {
        pub mod qu_x {
            // @generated
            // This file is @generated by prost-build.
            #[allow(clippy::derive_partial_eq_without_eq)]
            #[derive(Clone, PartialEq, ::prost::Message)]
            pub struct Bar {
                #[prost(string, tag = "1")]
                pub name: ::prost::alloc::string::String,
                #[prost(message, optional, tag = "2")]
                pub foo:
                    ::core::option::Option<::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::Foo>,
                #[prost(message, optional, tag = "3")]
                pub nested_foo: ::core::option::Option<
                    ::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::foo::NestedFoo,
                >,
            }
            /// Nested message and enum types in `Bar`.
            pub mod bar {
                #[allow(clippy::derive_partial_eq_without_eq)]
                #[derive(Clone, PartialEq, ::prost::Message)]
                pub struct Baz {
                    #[prost(string, tag = "4")]
                    pub name: ::prost::alloc::string::String,
                }
            }
            // @@protoc_insertion_point(module)
        }
    }
}
pub mod qaz {
    pub mod qu_x {
        // @generated
        // This file is @generated by prost-build.
        #[allow(clippy::derive_partial_eq_without_eq)]
        #[derive(Clone, PartialEq, ::prost::Message)]
        pub struct Bar {
            #[prost(string, tag = "1")]
            pub name: ::prost::alloc::string::String,
            #[prost(message, optional, tag = "2")]
            pub foo: ::core::option::Option<::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::Foo>,
            #[prost(message, optional, tag = "3")]
            pub nested_foo: ::core::option::Option<
                ::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::foo::NestedFoo,
            >,
        }
        /// Nested message and enum types in `Bar`.
        pub mod bar {
            #[allow(clippy::derive_partial_eq_without_eq)]
            #[derive(Clone, PartialEq, ::prost::Message)]
            pub struct Baz {
                #[prost(string, tag = "4")]
                pub name: ::prost::alloc::string::String,
            }
        }
        // @@protoc_insertion_point(module)
    }
}
pub mod qu_x {
    // @generated
    // This file is @generated by prost-build.
    #[allow(clippy::derive_partial_eq_without_eq)]
    #[derive(Clone, PartialEq, ::prost::Message)]
    pub struct Bar {
        #[prost(string, tag = "1")]
        pub name: ::prost::alloc::string::String,
        #[prost(message, optional, tag = "2")]
        pub foo: ::core::option::Option<::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::Foo>,
        #[prost(message, optional, tag = "3")]
        pub nested_foo: ::core::option::Option<
            ::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::foo::NestedFoo,
        >,
    }
    /// Nested message and enum types in `Bar`.
    pub mod bar {
        #[allow(clippy::derive_partial_eq_without_eq)]
        #[derive(Clone, PartialEq, ::prost::Message)]
        pub struct Baz {
            #[prost(string, tag = "4")]
            pub name: ::prost::alloc::string::String,
        }
    }
    // @@protoc_insertion_point(module)
}

```

Now:
```rust
// @generated

pub use foo_proto;

pub mod b_ar {
    pub mod b_az {
        pub mod qaz {
            pub mod qu_x {
                // @generated
                // This file is @generated by prost-build.
                #[allow(clippy::derive_partial_eq_without_eq)]
                #[derive(Clone, PartialEq, ::prost::Message)]
                pub struct Bar {
                    #[prost(string, tag = "1")]
                    pub name: ::prost::alloc::string::String,
                    #[prost(message, optional, tag = "2")]
                    pub foo: ::core::option::Option<
                        ::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::Foo,
                    >,
                    #[prost(message, optional, tag = "3")]
                    pub nested_foo: ::core::option::Option<
                        ::foo_proto::foo::quu_x::co_rg_e::grault::ga_rply::foo::NestedFoo,
                    >,
                }
                /// Nested message and enum types in `Bar`.
                pub mod bar {
                    #[allow(clippy::derive_partial_eq_without_eq)]
                    #[derive(Clone, PartialEq, ::prost::Message)]
                    pub struct Baz {
                        #[prost(string, tag = "4")]
                        pub name: ::prost::alloc::string::String,
                    }
                }
                // @@protoc_insertion_point(module)
            }
        }
    }
}
```

## Including transitive depenencies.

Because the proto rules rely on aspects, we are compiling targets which may not be directly accessible as a label in Bazel. Out of convenience, the original implementation automatically included all transitive dependencies to the `rust_library` or `rust_binary` that depended on the proto crate. 

As an example, let's say we have the following hierarchy:
```
my_crate -> a.proto
a.proto -> b.proto
b.proto -> c.proto
```
The previous setup made it possible for `my_crate` to directly import `a_proto`, `b_proto`, and `c_proto`.

This is problematic though because it leads to significant crate dependency bloat. `my_crate` may only use `a_proto` directly but it now also depends on the unused crates `b_proto` and `c_proto`.

This PR makes every generated crate re-export its direct dependencies. So, for the same example we now can import via those reexports:
```rust
//! my_crate
use a_proto::A;
use a_proto::b_proto::B;
use a_proto::b_proto::c_proto::C;
```

This means that you can also add the rustc flag `-Dunused_crate_dependencies` to ensure you're not depending on crates which are unused. In our monorepo we've had to disable this flag for crates depending on protos.

If you need to depend on an intermediate crate directly, you can just add a `rust_prost_library` definition and since it's all generated with aspects, it will be the identical underlying crate.

Finally, if you really want the extra crates included, you can set the flag `include_transitive_deps` on the prost toolchain and that will bring back the previous behavior.

### CODE REVIEW

File: proto/prost/private/tests/transitive_dependencies/transition.bzl [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+def _extra_toolchain_transition_impl(_settings, attr):
+    return {
+        "//command_line_option:extra_toolchains": [
+            attr.toolchain,
+        ],
+    }
+
+_extra_toolchain_transition = transition(
+    implementation = _extra_toolchain_transition_impl,
+    inputs = [],
```
Comment (UebelAndre) [inline proto/prost/private/tests/transitive_dependencies/transition.bzl:14]: It seems like RBE needs previously defined extra toolchains? Might want to append your new toolchain here.

File: proto/prost/private/tests/transitive_dependencies/transition.bzl [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    outputs = ["//command_line_option:extra_toolchains"],
+)
+
+def _extra_toolchain_wrapper_impl(ctx):
+    return [
+        ctx.attr.dep[DefaultInfo],
+        ctx.attr.dep[rust_common.crate_group_info],
+    ]
+
+extra_toolchain_wrapper = rule(
```
Comment (UebelAndre) [inline proto/prost/private/tests/transitive_dependencies/transition.bzl:24]: nit: can you fill out docs?

**Review by UebelAndre — CHANGES_REQUESTED** [review]
Thanks! Just a few questions!

File: proto/prost/private/tests/transitive_dependencies/a_test.rs [hunk 3]
```diff
@@ -1,11 +1,11 @@
 //! Tests transitive dependencies.
 
 use a_proto::a::A;
-use b_proto::a::b::B;
-use c_proto::a::b::c::C;
-use duration_proto::google::protobuf::Duration;
-use timestamp_proto::google::protobuf::Timestamp;
-use types_proto::Types;
+use a_proto::b_proto::a::b::B;
```
Comment (UebelAndre) [inline proto/prost/private/tests/transitive_dependencies/a_test.rs:4]: Can you also add a test that shows transitive dependencies become directly consumable as they did before if a `rust_prost_library` is made for it and added as a direct dependency to the the calling crate?

**Review by UebelAndre — CHANGES_REQUESTED** [review]

File: proto/prost/private/tests/transitive_dependencies/a_test.rs (hunk 3 above)
Comment (freeformstu) [reply proto/prost/private/tests/transitive_dependencies/a_test.rs:4]: I did that in the `transitve_deps_test.rs` which is beside this test.

File: proto/prost/private/tests/transitive_dependencies/transition.bzl (hunk 1 above)
Comment (freeformstu) [reply proto/prost/private/tests/transitive_dependencies/transition.bzl]: Thanks!

File: proto/prost/private/tests/transitive_dependencies/transition.bzl (hunk 2 above)
Comment (freeformstu) [reply proto/prost/private/tests/transitive_dependencies/transition.bzl:24]: Done

File: proto/prost/private/tests/transitive_dependencies/a_test.rs (hunk 3 above)
Comment (freeformstu) [reply proto/prost/private/tests/transitive_dependencies/a_test.rs]: I also updated the original test to take in b_proto as a direct dependency.

Comment (UebelAndre) [reply proto/prost/private/tests/transitive_dependencies/a_test.rs]: Thank you! That's exactly what I was thinking

**Review by UebelAndre — APPROVED** [review]

**Comment by bistack** [comment]
@freeformstu "a_proto::b_proto::c_proto::C" is too long when each x_proto is long.

As x_proto is a label name, it is more likely that there are multiple targets with same label name in a bigger monorepo.

So, we patched prost.bzl with a crate_name = label.package + label.name.

**Pull request merged**